File: tools/screenshot.js
/* Runs inside zenika/alpine-chrome:with-puppeteer (tools/screenshot.py copies it in): open a page
   of the built site at file:///site/<path>, and on a page with a graph wait for it to lay out and run
   the requested actions through window.graphmed (the hooks tools/site/static/graph.js exposes on a
   view page; home.js on the index exposes `open` alone); capture the viewport, or the whole page with
   spec.full. A view page takes every action; the index takes open=<view id>, sheet, graph, key=<key>
   and wait; any other page (an entity page) key= and wait. Any other action fails with its name, and
   what the page takes, before anything runs. Every page reports whether it is wider than the viewport,
   and where the focus is when it is not on the page itself. */
const puppeteer = require("/usr/src/app/node_modules/puppeteer");
const spec = JSON.parse(process.argv[2]);
const sleep = ms => new Promise(r => setTimeout(r, ms));
const errors = [];
const ANYWHERE = new Set(["wait", "key"]);   /* what a page without the graph takes */
const INDEX = new Set([...ANYWHERE, "open", "sheet", "graph"]);   /* what the index takes: its graph selects a guideline's box, and nothing else */
(async () => {
  const browser = await puppeteer.launch({ executablePath: "/usr/bin/chromium-browser", args: ["--no-sandbox", "--disable-gpu", "--hide-scrollbars"] });
  const page = await browser.newPage();
  page.on("pageerror", e => errors.push(String(e)));
  page.on("console", m => { if (m.type() === "error") errors.push(m.text()); });
  await page.setViewport({ width: spec.width, height: spec.height, deviceScaleFactor: spec.phone ? 2 : 1, isMobile: !!spec.phone, hasTouch: !!spec.phone });
  if (spec.dark) await page.emulateMediaFeatures([{ name: "prefers-color-scheme", value: "dark" }]);   /* before the load: the graph reads its colours once, when drawn */
  await page.goto(`file:///site/${spec.file}`, { waitUntil: "load" });
  /* a view page carries its graph's data; the index draws its graph from its own sheet */
  const kind = await page.evaluate(() => !document.getElementById("graph") ? "other" : document.getElementById("graph-data") ? "view" : "index");
  if (kind === "other") return other(browser, page);
  if (kind === "index") refuse(INDEX, "the index", "open=<view id>, sheet, graph, key=<key> and wait");
  await page.waitForFunction(() => window.graphmed && window.graphmed.cy.nodes().not(".folded").length > 0, { timeout: 20000 });
  await sleep(900);
  for (const [action, value] of spec.actions) {
    if (action === "wait") { await sleep(Number(value) || 500); continue; }
    if (action === "key") {
      const before = page.url();
      await press(page, value); await sleep(400);
      if (page.url().split("#")[0] !== before.split("#")[0])   /* a key that follows a link leaves the page: nothing of it is captured */
        throw new Error(`key=${value} followed a link to ${page.url().replace("file:///site/", "")}; a followed link cannot be captured, so stop the keys before it`);
      continue;
    }
    if (action === "graph" || action === "sheet") {   /* bring the graph or the details into view as a reader does */
      await page.evaluate(a => {
        /* on a phone a selection waits in a peek strip: `sheet` raises the panel by tapping it, `graph` lowers it again;
           where there is no strip (a wide screen, nothing selected), the page scrolls to the section instead */
        const peek = document.querySelector("#sheet .peek"), raised = document.querySelector("#sheet.raised");
        if (peek && peek.offsetParent !== null && (a === "sheet") !== !!raised) { peek.click(); return; }
        document.querySelector(a === "graph" ? ".graph-wrap" : "#sheet").scrollIntoView({ block: "start" });
      }, action);
      await sleep(400); continue;
    }
    if (action === "all") {   /* every patient group open, one tap at a time as a reader would */
      const ids = await page.evaluate(() => window.graphmed.cy.nodes("[type = 'junction']").map(j => j.id()));
      for (const id of ids) { await page.evaluate(id => window.graphmed.toggle(window.graphmed.cy.getElementById(id), true), id); await sleep(400); }
      await sleep(900); continue;
    }
    await page.evaluate((action, value) => {
      const g = window.graphmed;
      if (action === "toggle") g.toggle(g.cy.getElementById(value.indexOf("j:") === 0 ? value : "j:" + value));   /* a concept id, or a junction id in full (under an axis: j:<value>:<concept>) */
      else if (action === "by") g.by(value);
      else if (action === "fold") g.fold(g.cy.getElementById(value));
      else if (action === "reset") g.reset();
      else if (action === "open") g.open(value, false);
      else if (action === "section") g.section(value);
      else if (action === "search") g.search(value);
      else if (action === "facet") g.search(document.getElementById("search").value, value);
      else if (action === "step") for (let i = Math.abs(Number(value) || 1); i--;) g.step(Number(value) < 0 ? -1 : 1);   /* n steps through the matches, back when negative */
      else if (action === "chapters") document.getElementById("chapters-toggle").click();
      else if (action === "legend") document.getElementById("legend-toggle").click();   /* collapse or expand the legend */
      else if (action === "chapters-scroll") document.querySelector("#chapters .list").scrollTop = Number(value) || 0;
      else if (action === "fit") document.getElementById("fit").click();
      else throw new Error("unknown action " + action);
    }, action, value);
    await sleep(900);
  }
  await page.screenshot(spec.full ? { path: "/tmp/shot.png", fullPage: true } : { path: "/tmp/shot.png" });
  const shown = await page.evaluate(() => window.graphmed.cy.elements().not(".folded").length);
  /* what overlaps: every pair of shown nodes (with their labels) and edge labels whose boxes
     intersect by more than a pixel — the mechanical half of "nothing overlaps" (docs/publication.md §3) */
  const overlaps = await page.evaluate(() => {
    const cy = window.graphmed.cy, boxes = [];
    cy.nodes().not(".folded").forEach(n => { const b = n.boundingBox({ includeLabels: true, includeOverlays: false }); boxes.push({ id: n.id(), name: n.data("label") || n.id(), ...b }); });
    cy.edges().not(".folded").forEach(e => {
      if (!e.data("label")) return;
      const b = e.boundingBox({ includeEdges: false, includeLabels: true, includeOverlays: false });
      if (b.w > 0 && b.h > 0) boxes.push({ id: e.id(), name: "answer " + e.data("label"), ...b });
    });
    const pairs = [];
    for (let i = 0; i < boxes.length; i++) for (let j = i + 1; j < boxes.length; j++) {
      const a = boxes[i], b = boxes[j];
      const w = Math.min(a.x2, b.x2) - Math.max(a.x1, b.x1), h = Math.min(a.y2, b.y2) - Math.max(a.y1, b.y1);
      if (w > 1 && h > 1) pairs.push(`${a.name} × ${b.name} (${Math.round(w)}×${Math.round(h)})`);
    }
    /* and every edge drawn across a node it does not touch: a taxi edge as its three runs, any other
       as the straight line between its endpoints (a bezier bows a little, a segment is the least it covers) */
    const nodes = cy.nodes().not(".folded");
    const labels = boxes.filter(b => b.name.startsWith("answer "));
    function cuts(p, q, b) {   /* does segment p–q cross box b, shrunk by a pixel */
      const x1 = b.x1 + 1, y1 = b.y1 + 1, x2 = b.x2 - 1, y2 = b.y2 - 1;
      let t0 = 0, t1 = 1; const dx = q.x - p.x, dy = q.y - p.y;
      for (const [pp, qq] of [[-dx, p.x - x1], [dx, x2 - p.x], [-dy, p.y - y1], [dy, y2 - p.y]]) {
        if (pp === 0) { if (qq < 0) return false; continue; }
        const r = qq / pp;
        if (pp < 0) { if (r > t1) return false; if (r > t0) t0 = r; } else { if (r < t0) return false; if (r < t1) t1 = r; }
      }
      return t0 <= t1;
    }
    cy.edges().not(".folded").forEach(e => {
      const s = e.sourceEndpoint(), t = e.targetEndpoint(), runs = [];
      if (e.style("curve-style") === "taxi") { const turn = Number(e.data("turn")) || 24, x = turn < 0 ? t.x + turn : s.x + turn; runs.push([s, { x, y: s.y }], [{ x, y: s.y }, { x, y: t.y }], [{ x, y: t.y }, t]); }
      else runs.push([s, t]);
      const name = `edge ${e.source().data("label") || e.source().id()} → ${e.target().data("label") || e.target().id()}`;
      nodes.forEach(n => {
        if (n.same(e.source()) || n.same(e.target())) return;
        const b = n.boundingBox({ includeLabels: true, includeOverlays: false });
        if (runs.some(([p, q]) => cuts(p, q, b))) pairs.push(`${name} across ${n.data("label") || n.id()}`);
      });
      /* the answer before a target is written on the final run of every edge into it: not an obstacle for those */
      const own = new Set(e.target().incomers("edge").map(f => f.id()));
      labels.forEach(b => { if (!own.has(b.id) && runs.some(([p, q]) => cuts(p, q, b))) pairs.push(`${name} across ${b.name}`); });
    });
    return pairs;
  });
  console.log(`${shown} elements shown, ${overlaps.length} overlapping pairs; ${await across(page)}${await focus(page)}` + (errors.length ? `; page errors: ${errors.join(" | ")}` : ""));
  overlaps.slice(0, 40).forEach(p => console.log("  " + p));
  await browser.close();
})().catch(e => { console.error("screenshot failed: " + e.message + (errors.length ? "; page errors: " + errors.join(" | ") : "")); process.exit(1); });

/* an action the page does not take fails before anything runs, naming itself and what the page takes */
function refuse(takes, what, list) {
  const needs = spec.actions.find(([a]) => !takes.has(a));
  if (needs) throw new Error(`action ${needs[0]}${needs[1] ? "=" + needs[1] : ""} needs a view page; ${spec.file} is ${what}, which takes ${list}`);
}

/* a page without the graph: a view action fails, naming itself; otherwise the page settles, and the
   report is its page errors and whether it is wider than the viewport — the mechanical half of
   "the page fits a phone" (docs/publication.md) */
async function other(browser, page) {
  refuse(ANYWHERE, "a page without a graph", "key=<key> and wait");
  await page.evaluate(() => document.fonts.ready);
  await sleep(300);
  for (const [action, value] of spec.actions) {
    if (action === "wait") await sleep(Number(value) || 500);
    else if (action === "key") {
      const before = page.url();
      await press(page, value); await sleep(400);
      if (page.url().split("#")[0] !== before.split("#")[0])
        throw new Error(`key=${value} followed a link to ${page.url().replace("file:///site/", "")}; a followed link cannot be captured, so stop the keys before it`);
    }
  }
  await page.screenshot(spec.full ? { path: "/tmp/shot.png", fullPage: true } : { path: "/tmp/shot.png" });
  const tall = await page.evaluate(() => document.documentElement.scrollHeight);
  console.log(`no graph; ${await across(page)}, ${tall} px tall${await focus(page)}` + (errors.length ? `; page errors: ${errors.join(" | ")}` : ""));
  await browser.close();
}

/* key=<key>: a key pressed as a keyboard presses it — Tab, Enter, Escape, ArrowDown — with its modifiers joined by
   "+" (Shift+Tab); the way to check that a control is reached and works without a pointer */
async function press(page, value) {
  const keys = String(value || "").split("+").filter(Boolean);
  if (!keys.length) throw new Error("action key needs a key, e.g. key=Tab");
  const last = keys.pop();
  for (const k of keys) await page.keyboard.down(k);
  await page.keyboard.press(last);
  for (const k of keys.reverse()) await page.keyboard.up(k);
}

/* where the focus is, when it is not on the page itself: the element, its class, its words and whether it is
   pressed — what a key= run reached */
async function focus(page) {
  const at = await page.evaluate(() => {
    const el = document.activeElement;
    if (!el || el === document.body || el === document.documentElement) return "";
    const words = (el.getAttribute("aria-label") || el.textContent || "").replace(/\s+/g, " ").trim();
    const pressed = el.getAttribute("aria-pressed");
    return `${el.tagName.toLowerCase()}${el.className && typeof el.className === "string" ? "." + el.className.trim().split(/\s+/).join(".") : ""}`
      + (words ? ` "${words.length > 60 ? words.slice(0, 57) + "…" : words}"` : "") + (pressed ? ` (pressed ${pressed})` : "");
  });
  return at ? `; focus on ${at}` : "";
}

/* whether the page is wider than the viewport — the mechanical half of "the page fits a phone". Measured against
   the width asked for as well: a phone's layout viewport (innerWidth) widens to a page that overflows it */
async function across(page) {
  const [wide, inner] = await page.evaluate(() => [document.documentElement.scrollWidth, window.innerWidth]);
  const view = Math.min(inner, spec.width);
  return wide > view ? `overflows horizontally: ${wide} px wide at ${view}` : `fits ${view} px across`;
}

File: tools/screenshot.py
#!/usr/bin/env python3
"""Screenshot a page of the built site in a headless browser (the `screenshot` skill).

    uv run tools/screenshot.py pomgat-lv-1.0                       # a view: the folded start, 1280×900
    uv run tools/screenshot.py pomgat-lv-1.0 --phone               # 390×844 at device scale 2
    uv run tools/screenshot.py pomgat-lv-1.0 --dark                # the dark theme (prefers-color-scheme: dark)
    uv run tools/screenshot.py pomgat-lv-1.0 --do toggle=concepts/leberresektion \\
        --do open=statements/drainage-komplexe-leberresektion-optional --out /tmp/graph.med/screenshots/liver.png
    uv run tools/screenshot.py pomgat-lv-1.0 --phone --do open=statements/drainage-komplexe-leberresektion-optional --do graph
    uv run tools/screenshot.py statements/tap-block-mic-kolorektal --phone --full   # any page, the whole scrolled page
    uv run tools/screenshot.py /                                   # the index
    uv run tools/screenshot.py / --phone --do open=views/pomgat-lv-1.0   # the index, a guideline's box tapped

The page is a site path: a view id, an entity page (statements/<id>, concepts/<id>, axes/<id>,
sources/<id>, …), or / for the index; a trailing slash and index.html may be left out.

The sandbox has no browser and cannot download one, but it runs a Docker daemon and
image pulls pass the proxy: the page is rendered by Chromium inside a container
(zenika/alpine-chrome:with-puppeteer, which bundles Node and Puppeteer), driven by
tools/screenshot.js so that the layout can settle before the capture. The site is
built into a temporary directory with base path /site/ and copied into the container;
nothing is installed in the sandbox and nothing is mounted. Actions run in order
before the capture, and all but wait need a page with a graph (elsewhere the run fails naming the
action) — a view page takes every one, the index open=<view id> (select a guideline's box), sheet and
graph: by=<grouping> (choose the grouping: "" for the view's first, section for the
chapters, or an axis id),
toggle=<concept id> (fold or unfold that patient group; under an axis the junction id in full,
j:<value>:<concept id>),
fold=<question node id> (fold or unfold everything below that question, e.g.
q:j:concepts/leberresektion:population), open=<entity id> (deep link: unfold and select),
section=<number> (chapter filter), search=<text>, facet=<kind>, step=<n> (n steps through the
matches, back when negative), chapters (open the chapter panel), chapters-scroll=<px> (scroll its
list), all (every patient group open), fit (fit what is open), reset (the opening state), graph and
sheet (scroll the page to the graph or to the sheet: on a phone they stack), wait=<ms>. --full captures
the whole scrolled page instead of the viewport. --dark renders the page in the dark theme: the graph
reads its colours from the stylesheet once, when drawn, so the theme is emulated before the
page loads rather than switched by an action. On a page with a graph the runner prints how many
elements are shown and how many pairs of nodes and answers overlap — the mechanical half of "nothing
overlaps"; on every page, whether it is wider than the viewport — the mechanical half of "the page
fits a phone" — and any page error.
The container name and the output directory default to the current branch, so that
sessions working in parallel (one git worktree each, ADR-0002 and the
process-work-package skill) on the one Docker daemon do not remove each other's container or PNG.
"""

from __future__ import annotations

import argparse
import json
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
IMAGE = "zenika/alpine-chrome:with-puppeteer"
DRIVER = Path(__file__).resolve().parent / "screenshot.js"


def run(*cmd, **kw):
    return subprocess.run(cmd, check=True, text=True, capture_output=True, **kw)


def branch_slug() -> str:
    """The current branch as a name Docker and a path accept; 'main' when there is none."""
    try:
        name = run("git", "-C", str(ROOT), "rev-parse", "--abbrev-ref", "HEAD").stdout.strip() or "main"
    except (subprocess.CalledProcessError, FileNotFoundError):
        name = "main"
    return re.sub(r"[^A-Za-z0-9_.-]+", "-", name).strip("-.") or "main"


def main(argv=None) -> int:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("page", help="site path: a view id (pomgat-lv-1.0), an entity page (statements/<id>), or / for the index")
    ap.add_argument("--out", type=Path, default=None, help="PNG to write (default /tmp/graph.med/screenshots/<branch>/<page>.png)")
    ap.add_argument("--size", default="1280x900", help="viewport WxH (default 1280x900)")
    ap.add_argument("--phone", action="store_true", help="390x844 at device scale 2, touch")
    ap.add_argument("--full", action="store_true", help="capture the whole scrolled page, not only the viewport")
    ap.add_argument("--dark", action="store_true", help="the dark theme (emulates prefers-color-scheme: dark before the page loads)")
    ap.add_argument("--do", action="append", default=[], metavar="ACTION", help="an action before the capture; repeatable, in order")
    ap.add_argument("--name", default=None, help="container name (default shot-<branch>)")
    ap.add_argument("--preview", type=int, default=None, metavar="N", help="build as the preview of pull request N (the strip above the header)")
    args = ap.parse_args(argv)

    if shutil.which("docker") is None:
        print("error: docker is not available; the screenshot skill needs the sandbox's Docker daemon", file=sys.stderr)
        return 1
    branch = branch_slug()
    path = re.sub(r"(^|/)index\.html$", "", args.page.strip()).strip("/")
    file = f"{path}/index.html" if path else "index.html"
    stem = path.replace("/", "-") or "index"
    out = args.out or Path("/tmp/graph.med/screenshots") / branch / f"{stem}{'-phone' if args.phone else ''}{'-dark' if args.dark else ''}{'-full' if args.full else ''}.png"
    out.parent.mkdir(parents=True, exist_ok=True)
    args.name = args.name or f"shot-{branch}"
    width, height = (390, 844) if args.phone else map(int, args.size.lower().split("x"))
    spec = {"file": file, "width": width, "height": height, "phone": args.phone, "dark": args.dark, "full": args.full,
            "actions": [a.split("=", 1) if "=" in a else [a, ""] for a in args.do]}

    if not run("docker", "images", "-q", IMAGE).stdout.strip():
        print(f"pulling {IMAGE} (once per sandbox) …", file=sys.stderr)
        subprocess.run(["docker", "pull", "--quiet", IMAGE], check=True)

    with tempfile.TemporaryDirectory(prefix="graph.med-site-") as tmp:
        site = Path(tmp) / "site"
        build = [sys.executable, str(ROOT / "tools" / "build.py"), "--base", "/site/", "--out", str(site)]
        subprocess.run(build + (["--preview", str(args.preview)] if args.preview else []), check=True, capture_output=True)
        if ".." in Path(file).parts or not (site / file).is_file():
            print(f"error: no page {args.page!r} in the built site", file=sys.stderr)
            return 1
        subprocess.run(["docker", "rm", "-f", args.name], capture_output=True)
        run("docker", "create", "--name", args.name, "--entrypoint", "node", IMAGE, "/driver.js", json.dumps(spec))
        try:
            run("docker", "cp", str(site), f"{args.name}:/site")
            run("docker", "cp", str(DRIVER), f"{args.name}:/driver.js")
            started = subprocess.run(["docker", "start", "-a", args.name], text=True, capture_output=True)
            if started.returncode != 0:
                print(started.stdout + started.stderr, file=sys.stderr)
                return started.returncode
            run("docker", "cp", f"{args.name}:/tmp/shot.png", str(out))
        finally:
            subprocess.run(["docker", "rm", "-f", args.name], capture_output=True)
        info = started.stdout.strip()
    print(f"{out}  ({info})")
    return 0


if __name__ == "__main__":
    sys.exit(main())
